Secrets Import documentation (#25594)
hashicorp/vault: website/content/docs/import/index.mdx

---
layout: docs
page_title: Secrets import
description: Secrets import allows you to safely onboard secrets from external sources into Vault KV for management.
---

# Secrets import

<EnterpriseAlert product="vault" />

> @include 'alerts/beta.mdx'

Reviewer: Same comment here, we might need guidance to create an `alpha` or `experimental` alert.

Author: I've added an experimental that I'll push up soon to be reviewed.

Having collections of secrets across various external systems can present several challenges to an organization, such as
increased operational overhead, a greater risk of secrets sprawl, and more. Vault can offer a single source of truth (SSOT)
solution to managing these secrets, providing increased security and reducing that overhead, yet the effort required
to migrate all your secrets from multiple external sources into Vault can be steep. To aid in this onboarding process,
secrets import provides a mechanism to automate it. 

A secret that is imported from an external system is stored in Vault's KVv2 Secrets Engine. The import process is defined 
an import plan, which is an HCL file. Using HCL enables users to declaratively configure which of their secrets will be read
from an external source and where exactly they will be stored in Vault.

There are three main types of HCL blocks used to configure a secrets import plan. They are source blocks, the Vault destination

> block, and mapping blocks.

Reviewer: Should we loop in product or marketing here to validate the wording fits with their vision? It feels like a snippet they'd normally provide or would like to review.

## Vault destinations

Secrets that get imported are stored into a Vault KVv2 Secrets Engine mount. The destination will accept the mount path of
the desired KVv2 mount and optionally a namespace. This specifies the exact location to store the secrets in. A destination
is defined by an HCL block in the format of `destination_vault`.

### HCL syntax

#### Argument reference

The following arguments are supported:

> * `name` - (Required) The name of the Vault destination block used to internally reference this destination inside of a mapping.

Reviewer: Any reason we are not using the usual syntax of `name (string: <required>)`? Same question for all HCL fields.

Author: I was mimicking the Terraform provider docs style for how they list fields inside of resource blocks, since these aren't API fields. Not married to my idea but figured it makes sense to start with.

* `address` - (Optional) The address of the Vault server with the KVv2 Secrets Engine for which the secrets will be imported into.

> * `token` - (Optional) A token for authentication to the Vault server located at the specified address.

Reviewer: Should we document the address and token if they are always overridden? There is no reason for a Vault user to define these in their config file.

Author: I don't believe they are always overridden, I think they are only loaded from the environment if they are unset in the plan, hence why it's optional. Do you know if I am forgetting something?

> * `namespace` - (Optional) The name of the Vault namespace containing the specified KVv2 mount.

Reviewer: I'm not sure what is the typical Vault term for this but should it be the namespace "path". E.g. ns1/ns2. In this example to me the namespace name is ns2 but it wouldn't work if the parent namespace is omitted.

The field is optional, what is the default value?

> * `mount` - (Optional) The name of the KVv2 mount in Vault for which the secrets will be imported into.

Reviewer: I think this is the path, not the name.

Same comment here, if it is optional what is the default value?

#### Example

An example `destination_vault` block may look like the following:

```hcl
destination_vault {
  name = "my-dest-1"
  address = "http://127.0.0.1:8200/"
  token = "root"
  namespace = "ns-1"
  mount = "mount-1"
}

> ```

Reviewer: Same comment here, equal signs aligned & address + token should not be specified.

> An example `destination_vault` block may look like the following:
> 
> ```hcl
> destination_vault {
>   name = "my-dest-1"
>   address = "http://127.0.0.1:8200/"
>   token = "root"
>   namespace = "ns-1"
>   mount = "mount-1"
> }
> ```

Reviewer 2: ```suggestion
```

Moved above field list

## Sources

> Secrets can be imported from various external systems, called sources. The supported destinations are:

Reviewer: ```suggestion
Secrets can be imported from various external systems, called sources. The supported sources are:
```

* [GCP Secret Manager](/vault/docs/import/gcpsm)

A source requires the credentials necessary to read the external system's secrets and their values so that it may import
them into Vault. Either credentials to the system can be directly provided, or you can leverage Vault's capabilities to

> geneerate a dynamic secret for the importer to use if you have the corresponding secrets engine configured. A source is

Reviewer: ```suggestion
generate a dynamic secret for the importer to use if you have the corresponding secrets engine configured. A source is
```

defined by an HCL block in the format of `source_<external_system>`. For example, `source_gcp` defines a source block
for GCP Secret Manager.

### HCL syntax

Refer to a [specific source's documentation](#sources) for details on how to define all remaining arguments of a source
block for that external system type.

#### Example

A `source_<external_system>` block may look like the following, using GCP as an example:

```hcl
source_gcp {
  name = "my-gcp-source-1"
  credentials = "@/path/to/service-account-key.json"
}
```

## Mappings

> Mappings blocks are what glue together a source block to a destination block. The supported mapping types are:
> * [mapping_passthrough](/vault/docs/import/mappings#passthrough)
> * [mapping_metadata](/vault/docs/import/mappings#metadata)
> * [mapping_regex](/vault/docs/import/mappings#regex)
> 
> A mapping is where filtering logic is defined to control which secrets in the external system will be imported and which will
> be ignored. The definition of the rule depends on the type of mapping block that is being configured. Mapping blocks are
> applied in order of their given priority.
> 
> Once a secret being imported matches a single mapping's rule, it cannot match to another mapping again. A mapping is defined
> by an HCL block in the format of `mapping_<filter_type>`. For example, `mapping_regex` defines a mapping block to pass
> all secrets through whose secret name passes the given regular expression.

Reviewer 2: ```suggestion
Mappings glue the source and destination together and filter the migrated data
to control what is imported and what is ignored. Vault currently supports the
following mapping methods:

- [mapping_passthrough](/vault/docs/import/mappings#passthrough)
- [mapping_metadata](/vault/docs/import/mappings#metadata)
- [mapping_regex](/vault/docs/import/mappings#regex)
```

### HCL syntax

> See the [mapping documentation](/vault/docs/import/mappings) to see more info regarding specific mapping types and
> their arguments.
> 
> #### Example
> 
> A `mapping_<filter_type>` block may look like the following, using regex as an example:
> 
> ```hcl
> mapping_regex {
>   name = "my-map-1"
>   source = "my-gcp-source-1"
>   destination = "my-dest-1"
>   priority = 1
>   expression = "^database/.*$"
> }
> ```

Reviewer 2: ````suggestion
Mapping blocks start with `mapping_<filter_type>` and include a source name,
destination name, and any transformations or filters that apply. For example:

```hcl
mapping_regex {
  name        = "my-map-1"
  source      = "my-gcp-source-1"
  destination = "my-dest-1"
  priority    = 1
  expression  = "^database/.*$"
}
```

- `name (string: <required>)` - A unique name for the mapping block.

- `source (string: <required>)` - The name of a previously defined source block
  **from** which the data should be read.

- `destination (string: <required>)` - The name of a previously defined
  destination block **to** which the data should be written.

- `priority (integer: <required>)` - The order in which Vault should apply the
  mapping block during the import process. The lower the number, the higher the
  priority. For example, a mapping with priority 1 executes before a mapping
  with priority 2.

Depending on the filter type, additional fields may be required/possible. Refer
to the filter documentation for the different mappings for a list of
mapping-specific fields.

<Tip title="Priority matters">

  Vault applies mapping definitions in priority order and a given secret only
  matches to the first mapping that applies. Once Vault imports a secret with a
  particular mapping, subsequent reads from the same source will ignore that
  secret.

</Tip>
````
